### DISCUSSION TITLE
Testing APIs where the final response is from a potentially long delayed Webhook?

### DISCUSSION THREAD

**Discussion by rberger** [artilleryio/artillery#1374] [Artillery]
Is there a way to use Artillery to do end-to-end tests of an api that the http request will return 200 immediately, but the actual result that is needed to evaluate the success or failure of the API request comes back as a webhook request? And that request could be minutes or even longer.

I’m looking for a api testing / load testing platform like Jmeter, Artillery, Gatling, etc that could consider  a “test” to consist of:
* Make a HTTP REST request to my service under test that responds immediately with 200 (assuming proper arguments) starting a session of a test instance cycle
* The API testing platform can then accept an asynchronous webhook callback as the actual response to the initial request for a session
* Validates that webhook response is correct and that completes the actual test for that session
* Generates the stats for the session
* Can do that for lots of sessions in a load testing scenario
* Able to handle the webhook response in a way that doesn’t impact the load testing stats
![API with Webhook response SequenceDiagram1](https://user-images.githubusercontent.com/25268/158295327-1569ff63-a638-4283-9109-074a8db9fde9.jpg)

**Comment by hassy** [comment]
Hi @rberger :wave: that's a very interesting use-case!

Artillery does not have support for acting as a webhook target right now, which I think is what we'd need to support this use-case.

How do you correlate webhook calls to original requests to the API in your case? and how complex is the logic to validate the response?

> Able to handle the webhook response in a way that doesn’t impact the load testing stats

Do you mean that you'd only be interested in stats for the calls to the API endpoint that kick-off the long running task?

**Reply by rberger** [reply]
Yes, I live in interesting times :-)

> Artillery does not have support for acting as a webhook target right now, which I think is what we'd need to support this use-case.

I suspect that is the crux of the matter. Wondering if there are any work arounds that would make it possible to still implement this use case. Could the extension mechanism be used possibly? Or any other  possible mechanisms to get such a service?

> How do you correlate webhook calls to original requests to the API in your case?

There would be an id that is effectively the event or session id that is shared between the request and the webhook.

>  and how complex is the logic to validate the response?

I believe that the response will include an XML or JSON file (we have both use cases) that can be inspected for specific values to determine if the response should be considered pass or fail.

> Do you mean that you'd only be interested in stats for the calls to the API endpoint that kick-off the long running task?

We're interested in stats of the total session time, i.e. from when the initial request is made until the webhook is received

The main thrust of my question: `Able to handle the webhook response in a way that doesn’t impact the load testing stats` is more about the need that the webhook target would have to scale up based on load the same way the test generator is scaling out so as to not become the bottleneck in measuring the throughput/load
